common/metrics/events_scope.go

```diff
@@ -0,0 +1,175 @@
+	m := MetricDefs[e.serviceIdx][counter]
+
+	e.provider.Counter(m.metricName.String(), &MetricOptions{
+		Unit: event.Unit(m.unit),
+	}).Record(int64(delta), scopeDefToTags(s)...)
+
+	if !m.metricRollupName.Empty() {
+		e.provider.Counter(m.metricRollupName.String(), &event.MetricOptions{
+			Unit: event.Unit(m.unit),
+		}).Record(int64(delta), scopeDefToTags(s)...)
+	}
+}
+
+// StartTimer starts a timer for the given metric name.
+// Time will be recorded when stopwatch is stopped.
+func (e eventsScope) StartTimer(timer int) Stopwatch {
+	return &eventsStopwatch{
+		recordFunc: func(d time.Duration) {
+			s := ScopeDefs[e.serviceIdx][e.scopeId]
+			m := MetricDefs[e.serviceIdx][timer]
+
+			e.provider.Timer(m.metricName.String(), nil).Record(d, scopeDefToTags(s)...)
+
+			if !m.metricRollupName.Empty() {
+				e.provider.Timer(m.metricRollupName.String(), nil).Record(d, scopeDefToTags(s)...)
+			}
+		},
+		start: time.Now(),
+	}
+}
+
+// RecordTimer records a timer for the given metric name
+func (e eventsScope) RecordTimer(timer int, d time.Duration) {
+	s := ScopeDefs[e.serviceIdx][e.scopeId]
+	m := MetricDefs[e.serviceIdx][timer]
+
+	e.provider.Timer(m.metricName.String(), &MetricOptions{
+		Unit: event.Unit(m.unit),
+	}).Record(d, scopeDefToTags(s)...)
+
+	if !m.metricRollupName.Empty() {
+		e.provider.Timer(m.metricRollupName.String(), &event.MetricOptions{
+			Unit: event.Unit(m.unit),
+		}).Record(d, scopeDefToTags(s)...)
+	}
+}
+
+// RecordDistribution records a distribution (wrapper on top of timer) for the given
+// metric name
+func (e eventsScope) RecordDistribution(id int, d int) {
+	s := ScopeDefs[e.serviceIdx][e.scopeId]
+	m := MetricDefs[e.serviceIdx][id]
+
+	e.provider.Histogram(m.metricName.String(), &MetricOptions{
+		Unit: event.Unit(m.unit),
+	}).Record(int64(d), scopeDefToTags(s)...)
+
+	if !m.metricRollupName.Empty() {
+		e.provider.Histogram(m.metricRollupName.String(), &event.MetricOptions{
+			Unit: event.Unit(m.unit),
+		}).Record(int64(d), scopeDefToTags(s)...)
+	}
+}
+
+// UpdateGauge reports Gauge type absolute value metric
+func (e eventsScope) UpdateGauge(gauge int, value float64) {
+	s := ScopeDefs[e.serviceIdx][e.scopeId]
+	m := MetricDefs[e.serviceIdx][gauge]
+
+	e.provider.Gauge(m.metricName.String(), &MetricOptions{
+		Unit: event.Unit(m.unit),
+	}).Record(value, scopeDefToTags(s)...)
+
+	if !m.metricRollupName.Empty() {
+		e.provider.Gauge(m.metricRollupName.String(), &event.MetricOptions{
+			Unit: event.Unit(m.unit),
+		}).Record(value, scopeDefToTags(s)...)
+	}
+}
+
+// Tagged returns an internal scope that can be used to add additional
+// information to metrics
+func (e eventsScope) Tagged(tags ...Tag) Scope {
+	return newEventsScope(e.provider.WithTags(tags...), e.serviceIdx, e.scopeId)
+}
+
+func scopeDefToTags(s scopeDefinition) []Tag {
+	t := make([]Tag, 0, len(s.tags)+1)
+
+	t = append(t, OperationTag(s.operation))
+
+	for k, v := range s.tags {
+		t = append(t, StringTag(k, v))
+	}
+
+	return t
+}
```

Comment:
might have perf issue here? Is scopeDefinition immutable? may need to cache the tags for each scopeDef.

Reply:
that is a good question, the struct is passed by value here, so in the context of this function changes won't be propagated.

I see the concern and this could be cached as after the first conversation from map to tags.